### PR TITLE
Create verify_blob_sidecar_inclusion_proof()

### PR DESCRIPTION
**Pull request by EchoAlice**
WIP

### CODE REVIEW

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (ralexstokes) [inline ethereum-consensus/src/deneb/blob_sidecar.rs:42]:
```suggestion
            .map(|bytes| Node::try_from(bytes.as_slice()).expect("correct size input"))
```

Comment (ralexstokes) [inline ethereum-consensus/src/deneb/blob_sidecar.rs:42]: nitpick
```suggestion
            .map(|bytes| Node::try_from(bytes.as_ref()).unwrap())
```

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (ralexstokes) [inline ethereum-consensus/src/deneb/blob_sidecar.rs:44]: move this to be a `pub const` in this file plz

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (ralexstokes) [inline ethereum-consensus/src/deneb/blob_sidecar.rs:45]:
```suggestion
```

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (ralexstokes) [inline ethereum-consensus/src/deneb/blob_sidecar.rs:35]: lets just define index once, this may change a bit after I land the ssz-rs updates
```suggestion
        let index = 4;
```

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (ralexstokes) [inline ethereum-consensus/src/deneb/blob_sidecar.rs:33]: let's use the crate-wide error for this, and make a new error type for this specific set of failure modes

imports up above:
```rust
use crate::Error;
```

```suggestion
    pub fn verify_blob_sidecar_inclusion_proof(&mut self) -> Error {
```

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (ralexstokes) [inline ethereum-consensus/src/deneb/blob_sidecar.rs:38]: w/ error changes then:
```suggestion
        let leaf = &self.kzg_commitment.hash_tree_root()?;
```

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (ralexstokes) [inline ethereum-consensus/src/deneb/blob_sidecar.rs:48]: I actually think I'll update this API some and include a bespoke error type, rather than just `bool`

we can make the changes once I land this and the g-index updates to `ssz-rs`

**Review by ralexstokes — COMMENTED** [review]
great start! left some comments, and this otherwise will be blocked on some `ssz-rs` updates ill try to get to soon(tm)

**Comment by ralexstokes** [comment]
@EchoAlice FYI updated the `is_valid_merkle_branch` API, so if you rebase from `main`, you can simplify your use of it here; check out that PR to see other places it changed for an example

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (EchoAlice) [reply ethereum-consensus/src/deneb/blob_sidecar.rs:42]: That's why they pay you the big bucks

**Review by ralexstokes — COMMENTED** [review]
nice updates! we need to implement the logic to compute the g_index (or could compute in pyspec and just copy over as a constant

**Comment by EchoAlice** [comment]
> nice updates! we need to implement the logic to compute the g_index (or could compute in pyspec and just copy over as a constant

Which option makes more sense to you?

**Comment by ralexstokes** [comment]
unblocked on the generalized index computation with #324

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (ralexstokes) [inline ethereum-consensus/src/deneb/blob_sidecar.rs:3]: you want to use the generic type here, not a particular preset

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (ralexstokes) [inline ethereum-consensus/src/deneb/blob_sidecar.rs:55]: just following the spec to the letter....
```suggestion
fn get_subtree_index(i: GeneralizedIndex) -> usize {
```

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (ralexstokes) [inline ethereum-consensus/src/deneb/blob_sidecar.rs:56]: the spec avoids floating point arithmetic for a number of reasons, one of them being determinism of execution

I'd just use this instead https://doc.rust-lang.org/std/primitive.usize.html#method.checked_ilog2

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (ralexstokes) [inline ethereum-consensus/src/deneb/blob_sidecar.rs:57]: just a note that `i as i32` would truncate the number, tossing away the upper half

Comment (ralexstokes) [inline ethereum-consensus/src/deneb/blob_sidecar.rs:57]: you should just be able to pipe the floor log 2 value into this

https://doc.rust-lang.org/std/primitive.usize.html#method.pow

**Review by ralexstokes — CHANGES_REQUESTED** [review]
nice updates! left a few more comments

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (EchoAlice) [reply ethereum-consensus/src/deneb/blob_sidecar.rs:57]: Yep.  I *assumed* we wouldn't be using giant numbers.  But understand that's probably bad form

Comment (ralexstokes) [reply ethereum-consensus/src/deneb/blob_sidecar.rs:57]: well its more that you don't know

your input is a 64-bit uint, so you should program assuming that

Comment (ralexstokes) [reply ethereum-consensus/src/deneb/blob_sidecar.rs:57]: types are useful *because* they communicate contracts

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (EchoAlice) [reply ethereum-consensus/src/deneb/blob_sidecar.rs:3]: i'm not sure how to import a generic BeaconBlockBody

Comment (ralexstokes) [reply ethereum-consensus/src/deneb/blob_sidecar.rs:3]:
```suggestion
        beacon_block::BeaconBlockBody,
```

you'll need to provide the const generic arguments which is why I was suggesting just having this be a free-standing function, rather than having types all over the place

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (ralexstokes) [inline ethereum-consensus/src/deneb/blob_sidecar.rs:55]: we should prefer to return an error, rather than panicking, which `expect` would do here if `i == 0`

instead, we can do `unwrap_or` and return this error: https://github.com/ralexstokes/ssz-rs/blob/6329986d14e71a98a35cc9323f88cbef0332b6dd/ssz-rs/src/merkleization/mod.rs#L38

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (ralexstokes) [inline ethereum-consensus/src/deneb/blob_sidecar.rs:34]: better to take `&mut` here so we don't run around cloning this data all the time

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (EchoAlice) [reply ethereum-consensus/src/deneb/blob_sidecar.rs:3]: I'm not sure how to provide the const generic arguments.  When i add the const generics to the appropriate function signatures, i get this error:

```    Checking ethereum-consensus v0.1.1 (/home/echo/Desktop/Ethereum/ethereum-consensus/ethereum-consensus)
error[E0282]: type annotations needed
  --> ethereum-consensus/src/deneb/blob_sidecar.rs:81:5
   |
81 |     BeaconBlockBody::generalized_index(path)
   |     ^^^^^^^^^^^^^^^ cannot infer the value of const parameter `MAX_PROPOSER_SLASHINGS` declared on the struct `BeaconBlockBody`
```

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (EchoAlice) [reply ethereum-consensus/src/deneb/blob_sidecar.rs:117]: `.unwrap_or()` expects a `u32` as input instead of an error.  I'm going explicitly handle the returned value of `.checked_ilog2()` with a `match` statement.  Thoughts?

Comment (ralexstokes) [reply ethereum-consensus/src/deneb/blob_sidecar.rs]: sorry I meant this: https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (ralexstokes) [inline ethereum-consensus/src/deneb/blob_sidecar.rs:116]:
```suggestion
```

Comment (EchoAlice) [reply ethereum-consensus/src/deneb/blob_sidecar.rs:116]: Oops

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (ralexstokes) [inline ethereum-consensus/src/deneb/blob_sidecar.rs:31]:
```suggestion
}

pub fn verify_blob_sidecar_inclusion_proof<
```

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (ralexstokes) [inline ethereum-consensus/src/deneb/blob_sidecar.rs:52]: ok, one last thing I think!

let's mirror [the spec](https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/p2p-interface.md#verify_blob_sidecar_inclusion_proof) and take a `BlobSidecar` as an input!

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (ralexstokes) [inline ethereum-consensus/src/deneb/blob_sidecar.rs:96]: either inline this function, or let's keep the typing
```suggestion
    i: BlobIndex,
```

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (ralexstokes) [inline ethereum-consensus/src/deneb/blob_sidecar.rs:77]: nit but I think this is easier to read, rather than the dense Ok() and ? usage

```suggestion
    is_valid_merkle_branch(leaf, branch, depth, subtree_index, root).map_error(Into::into)
```

**Review by ralexstokes — CHANGES_REQUESTED** [review]
one more round then we can merge :)

File: ethereum-consensus/src/deneb/blob_sidecar.rs
Comment (EchoAlice) [reply ethereum-consensus/src/deneb/blob_sidecar.rs:77]: i assume you meant `.map_err(Into::into)`, right?

**Comment by EchoAlice** [comment]
> perfect! thanks for all the back and forth :)

Thanks for all of the guidance!

**Pull request merged**
